native/dlarft.go

```diff
@@ -0,0 +1,137 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack"
+)
+
+// Dlarft forms the triangular factor t of a block reflector, storing the answer
+// in t.
+//  H = 1 - V * T * V^T if store == lapack.Columnwise
+//  H = 1 - V^T * T * V if store == lapack.Rowwise
+// H is defined by a product of the elementary reflectors where
+//  H = H_1 * H_2 * ... * H_k if direct == lapack.Forward
+//  H = H_k * H_k-1 * ... * H_1 if direct == lapack.Backward
+//
+// t is a k x k triangular matrix. t is upper triangular if direct = lapack.Forward
+// and lower triangular otherwise.
+//
+// store describes the storage of the elementary reflectors in v. Please see
+// Dlarfb for a description of layout.
+//
+// tau contains the scalar factor of the elementary reflectors h.
+func (Implementation) Dlarft(direct lapack.Direct, store lapack.StoreV, n, k int,
+	v []float64, ldv int, tau []float64, t []float64, ldt int) {
+	if n == 0 {
+		return
+	}
+	if n < 0 || k < 0 {
+		panic(negDimension)
+	}
+	if direct != lapack.Forward && direct != lapack.Backward {
+		panic(badDirect)
+	}
+	if store != lapack.Rowwise && store != lapack.Columnwise {
+		panic(badStore)
+	}
+	if len(tau) < k {
+		panic(badTau)
+	}
+	checkMatrix(k, k, t, ldt)
+	bi := blas64.Implementation()
+	// TODO(btracey): There are a number of minor obvious loop optimizations here.
+	// TODO(btracey): It may be possible to rearrange some of the code so that
+	// index of 1 is more common in the Dgemv.
+	if direct == lapack.Forward {
+		prevlastv := n - 1
+		for i := 0; i < k; i++ {
+			prevlastv = max(i, prevlastv)
+			if tau[i] == 0 {
+				for j := 0; j <= i; j++ {
+					t[j*ldt+i] = 0
+				}
+				continue
+			}
+			var lastv int
+			if store == lapack.Columnwise {
+				// skip trailing zeros
+				for lastv = n - 1; lastv >= i+1; lastv-- {
+					if v[lastv*ldv+i] != 0 {
+						break
+					}
+				}
+				for j := 0; j < i; j++ {
```

Comment:
I'm not sure what you mean. This looks like lines 222-224 of http://www.netlib.org/lapack/explore-html/dd/daa/dlarft_8f_source.html